test/tests/functional/pbs_node_rampdown_cset.py

```diff
@@ -0,0 +1,156 @@
+# License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Commercial License Information:
+#
+# PBS Pro is commercially licensed software that shares a common core with
+# the OpenPBS software.  For a copy of the commercial license terms and
+# conditions, go to: (http://www.pbspro.com/agreement.html) or contact the
+# Altair Legal Department.
+#
+# Altair's dual-license business model allows companies, individuals, and
+# organizations to create proprietary derivative works of OpenPBS and
+# distribute them - whether embedded or bundled with other software -
+# under a commercial license agreement.
+#
+# Use of Altair's trademarks, including but not limited to "PBS™",
+# "OpenPBS®", "PBS Professional®", and "PBS Pro™" and Altair's logos is
+# subject to Altair's trademark licensing policies.
+
+from tests.functional import *
+
+
+@requirements(num_moms=2)
+class TestPbsNodeRampDownCset(TestFunctional):
+
+    """
+    This tests the Node Rampdown Feature on a cluster where the
+    sister mom is a cgroup cpuset system.
+    """
+
+    def setUp(self):
+
+        TestFunctional.setUp(self)
+        Job.dflt_attributes[ATTR_k] = 'oe'
+
+        # skip if there are less than one regular vnode and
+        # three from cpuset system (natural + 2 NUMA vnodes)
+        nodeinfo = self.server.status(NODE)
+        if len(nodeinfo) < 4:
+            self.skipTest("Not enough vnodes to run the test.")
+
+        # skip if there are no cpuset systems in the test cluster
+        no_csetmom = True
+        for mom in self.moms.values():
+            if mom.is_cpuset_mom():
+                no_csetmom = False
+        if no_csetmom:
+            self.skipTest("Skip on cluster without cgroup cpuset system.")
+
+        # Various node names
+        self.n0 = self.moms.values()[0].shortname
+        self.n1 = self.moms.values()[1].shortname
+        self.n2 = '%s[0]' % (self.n1,)
+        self.n3 = '%s[1]' % (self.n1,)
+
+        # skip if second mom has less than two NUMA vnodes
+        try:
+            self.server.expect(NODE, {'state': 'free'},
+                               id=self.n3, max_attempts=10)
+        except PtlExpectError:
+            self.skipTest("Second mom has less than 2 vnodes")
+
+        a = {'state': 'free', 'resources_available.ncpus': (GE, 1)}
+        self.server.expect(VNODE, {'state=free': 4}, op=EQ, count=True,
+                           max_attempts=10, interval=2)
+
+        self.pbs_release_nodes_cmd = os.path.join(
+            self.server.pbs_conf['PBS_EXEC'], 'bin', 'pbs_release_nodes')
+
+        # number of resource ncpus to request initially
+        ncpus = self.server.status(NODE, 'resources_available.ncpus',
+                                   id=self.n3)[0]
+        ncpus = int(ncpus['resources_available.ncpus'])
+        self.ncpus2 = ncpus / 2 - 1
```

Comment:
So if 'ncpus' is 4, initial self.ncpus2 = 4/2 -1 = 1, and if 'ncpus' is 8, initial self.ncpus2 = 8/2 - 1 = 3. So what's the significance of the '-1'?

Reply:
I have removed the -1.